All instructions lifted (Z80 arch)

samrussell opened the pull request.

samrussell committed:
Lifted more opcodes, minor bugfix in ColecoView

bugfix: add perform_get_address_size(self) to ColecoView
feature: lifted RST, conditional CALL, CPL, NEG, RES, SET

samrussell committed:
Implement RRC and RRCA

samrussell committed:
Implement EXX opcode

samrussell committed:
expressionify: remove shortcut on int at start

samrussell committed:
Implement DAA

samrussell committed:
Implement BCD flags

samrussell committed:
minor bugfix, s/SBC/SBB

samrussell committed:
Implement Z flag for POP

samrussell committed:
Fix POP flag implementation, might help EXX

samrussell committed:
Complete POP and EX flag handling

samrussell committed:
Implement OUT IN, flags XOR OR BIT

samrussell committed:
DI EI HALT IM OUTIDR SCF SRA SRL and ADC flags

Added opcodes DI EI HALT IM OUTI OUTD OTIR OTDR SCF SLA SRL
Added extra flag sets on SRA
Implemented overflow flag for ADC (copied from SBB implementation)

samrussell committed:
INI IND INIR INDR

samrussell committed:
Fix DJNZ, add RLD and RRD

samrussell committed:
Add LDD, LDDR

samrussell committed:
CPI|D|R and RETI|RETN

samrussell committed:
Tidy flag push/pop code

Goal is to pretend the F part of the AF register doesn't exist. Pushing
and popping in single bytes tidies the POP slightly, and moved to using
il.flag_bit() which gives same results as PUSHFQ/POPFQ in the built-in
x64 lifter

plafosse commented:
Very cool! Thanks for the PR!

xusheng6 commented:
I will be reviewing this

samrussell committed:
set coleco byte size to 2

samrussell commented:
Hi guys, any chance someone could take a look at this? I can break it into smaller pieces if that makes life easier

galenbwill commented:
I will prioritize reviewing it.

xusheng6 commented on Z80IL.py:
```diff
@@ -377,51 +482,323 @@ def gen_instr_il(addr, decoded, il):
+        temp0 = LLIL_TEMP(0)
+        il.append(il.intrinsic([ILRegister(il.arch, temp0)], "in", [il.reg(1, 'C')]))
+        # save to (HL)
+        il.append(il.store(1, il.reg(2, 'HL'), temp0))
```
This should be il.reg(1, temp0) instead of just temp0. temp0 is the index of the temp register, using it directly does not refer to the register. This also happens on line 757, 773, 792

xusheng6 commented on Z80IL.py:
```diff
@@ -377,51 +482,323 @@ def gen_instr_il(addr, decoded, il):
+        tmp = il.sub(2, il.reg(2, 'BC'), il.const(2, 1))
+        il.append(il.set_reg(2, 'BC', tmp))
+        il.append(il.set_flag('pv', il.compare_not_equal(2, il.reg(2, 'BC'), il.const(2, 0))))
+        il.append(il.if_expr(il.or_expr(1, il.flag('z'), il.neg_expr(1, il.flag('pv'))), label_done, label_loop))
```
neg_expr does arithemetic negation, not logical, so I believe the condition is wrong. Same pattern on line 544

xusheng6 commented on Z80IL.py:
```diff
@@ -377,51 +482,323 @@ def gen_instr_il(addr, decoded, il):
+        # if lower nybble > 9 OR H flag set then we diff by 0x06
+        label_add_6 = LowLevelILLabel()
+        label_dont_add_6 = LowLevelILLabel()
+        lower_nybble = il.xor_expr(1, il.const(1, 0x0F), il.reg(1, 'A'))
```
I think this should be an and instead of xor, in order to extract the bits

Also as a comment, for instructions like DAA, we usually (e.g., for x86) just lift it as an intrinsic rather than spelling out the entire computation. I am not against your doing it like this since z80 is simple enough, so it is practical to do it for all instructions, but I am a bit worried that it hurts readability than helps it. After all this is merely a suggestion, you do not have to change anything

xusheng6 commented on Z80IL.py:
```diff
@@ -500,45 +905,126 @@ def gen_instr_il(addr, decoded, il):
+                size = 1
+            ))
+            temp0 = il.expr(LowLevelILOperation.LLIL_REG, LLIL_TEMP(0), 1)
+            il.append(il.set_flag('s', il.test_bit(1, temp0, il.const(1, 1<<7))))
+            il.append(il.set_flag('z', il.test_bit(1, temp0, il.const(1, 1<<6))))
+            il.append(il.set_flag('h', il.test_bit(1, temp0, il.const(1, 1<<4))))
+            il.append(il.set_flag('pv', il.test_bit(1, temp0, il.const(1, 1<<2))))
+            il.append(il.set_flag('n', il.test_bit(1, temp0, il.const(1, 1<<1))))
+            il.append(il.set_flag('c', il.test_bit(1, temp0, il.const(1, 1))))
```
il.test_bit tests the bit at the given index, there is no need to do `1<<7` -- just pass in il.const(1, 7). Same pattern at line 681

xusheng6 commented on Z80IL.py:
```diff
@@ -377,51 +482,323 @@ def gen_instr_il(addr, decoded, il):
         # decrement B
         tmp = il.reg(1, 'B')
-        tmp = il.add(1, tmp, il.const(1,-1))
+        tmp = il.sub(1, tmp, il.const(1, 1), 'z')
```
This will incorrectly update the value of the z flag, which DJNZ does not modify. You could try something like this:

```
  result = il.sub(1, il.reg(1, 'B'), il.const(1, 1))
  il.append(il.set_reg(1, 'B', result))

  condition = il.compare_not_equal(
      1,
      il.reg(1, 'B'),
      il.const(1, 0)
  )
  il.append(il.if_expr(condition, label_loop, label_continue))
```

xusheng6 commented on Z80IL.py:
```diff
@@ -377,51 +482,323 @@ def gen_instr_il(addr, decoded, il):
+        # BC = BC - 1
+        tmp = il.sub(2, il.reg(2, 'BC'), il.const(2, 1))
+        il.append(il.set_reg(2, 'BC', tmp))
+        il.append(il.set_flag('pv', il.compare_equal(2, il.reg(2, 'BC'), il.const(2, 0))))
```
I think the condition here is inverted -- it should be il.compare_not_equal

xusheng6 reviewed (changes requested):
I found a few issues in the code. Once these are fixed, it should be good to go.

Sorry again for our late resoonse. I will respond promptly once you get these addressed!

plafosse commented:
Unfortunately looks like we're going to have to slip this one to the next release.

xusheng6 committed:
Use and, not xor, to extract the low nybble in DAA

xusheng6 committed:
Pass bit indices to il.test_bit(), not bit masks

il.test_bit(size, value, index) tests the bit at the given index, so the
1<<n masks were testing the wrong bits. Use plain indices matching the F
register layout (c=0, n=1, pv=2, h=4, z=6, s=7) in EX AF,AF' and POP AF,
and bit 0 / bit 7 for the carry out of the shifts and rotates. Also size
the EX AF,AF' tests as 1 byte, since F is a 1-byte register.

xusheng6 committed:
Store the temp register value, not its index, in INI/INIR/IND/INDR

il.store() takes an expression, but the raw LLIL_TEMP() index was being
passed. Wrap it in il.reg(1, temp0) as OP.IN already does.

xusheng6 committed:
Fix P/V for CPI and CPD

Per the manual P/V is set if BC-1 is not zero, and reset otherwise. This
also matches what CPIR/CPDR already do.

xusheng6 committed:
Fix CPIR/CPDR loop exit condition

neg_expr() is arithmetic negation, not a logical not, so the exit test
was wrong. Compare BC against zero directly instead: the loop ends when
A == (HL) (z set) or BC has reached zero.

xusheng6 committed:
Stop DJNZ from clobbering the z flag

DJNZ modifies no flags. Decrement B without a flag write and branch on a
direct comparison of B against zero.

xusheng6 committed:
Pass the I/O port to the in/out intrinsics as a value

The decoder types the (n) and (C) operands of IN/OUT as ADDR_DEREF /
REG_DEREF, since z80 syntax spells ports with parentheses. Running those
through operand_to_il() wrapped them in a load, so OUT (0xfe),A lifted as
out([0xfe].b, A) -- claiming a memory read at 0x00fe and making Binary
Ninja synthesize a data_fe variable and a cross reference there.

The in/out intrinsics declare the port as Type.int(1), so pass it
directly. INI/IND/OUTI/OUTD already did this.

xusheng6 committed:
Define the cszpv flag write type

'cszpv' was listed in flag_write_types but missing from
flags_written_by_flag_write_type, so it resolved to the empty list and
SLA/SRA/SRL wrote no flags at all. The declaration predates this branch,
but nothing used it until SRA moved from flags='c' to flags='cszpv'
here, which regressed SRA's carry.

Also add carry lifting for LLIL_LSL and LLIL_LSR so SLA and SRL compute
a carry, matching the existing LLIL_ASR rule.

xusheng6 commented:
Thanks a lot @samrussell!! I made the changes and merged the PR

samrussell commented:
Thanks Xusheng, didn't get a chance to implementing myself, glad to get it
all merged

On Thu, Aug 20, 2026, 17:48 xusheng ***@***.***> wrote:

> *xusheng6* left a comment (Vector35/Z80#10)
> <https://github.com/Vector35/Z80/pull/10#issuecomment-5358311318>
>
> Thanks a lot @samrussell <https://github.com/samrussell>!! I made the
> changes and merged the PR
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Vector35/Z80/pull/10?email_source=notifications&email_token=AALJUP7HYVNCFCBYHRYFBMT5K4MWJA5CNFSNUABFM5UWIORPF5TWS5BNNB2WEL2JONZXKZKDN5WW2ZLOOQXTKMZVHAZTCMJTGE4KM4TFMFZW63VHNVSW45DJN5XKKZLWMVXHJLDGN5XXIZLSL5RWY2LDNM#issuecomment-5358311318>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AALJUP4CDRBAZWDBG362AXD5K4MWJAVCNFSNUABFKJSXA33TNF2G64TZHMZDAMBWG43TSMZYHNEXG43VMU5TEMJWGQ4DMOBSHA2KC5QC>
> .
> Triage notifications, keep track of coding agent tasks and review pull
> requests on the go with GitHub Mobile for iOS
> <https://github.com/notifications/mobile/ios/AALJUP2CLYH3YYNCBYMKQM35K4MWJA5CNFSNUABFM5UWIORPF5TWS5BNNB2WEL2JONZXKZKDN5WW2ZLOOQXTKMZVHAZTCMJTGE4KM4TFMFZW63VHNVSW45DJN5XKKZLWMVXHJKTGN5XXIZLSL5UW64Y>
> and Android
> <https://github.com/notifications/mobile/android/AALJUP4JETPJYNC6VLXMLAD5K4MWJA5CNFSNUABFM5UWIORPF5TWS5BNNB2WEL2JONZXKZKDN5WW2ZLOOQXTKMZVHAZTCMJTGE4KM4TFMFZW63VHNVSW45DJN5XKKZLWMVXHJLTGN5XXIZLSL5QW4ZDSN5UWI>.
> Download it today!
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
